t/issue/1508.t

```diff
@@ -0,0 +1,165 @@
+
```

Comment:
As mentioned, and despite how it may look at first, this test script was heavily inspired from `t/needs.t`. 

It's basically a refactored version of it, with one notable functional difference: unlike  `t/needs.t`, this one also checks the run-time task options and params. 

So, my initial -joking- response would be : 

sure, please walk us through `t/needs.t` and I will do the same for this one :-)

Reply:
Joking aside, here's a brief run-through :

Just like in `t/needs.t`, the `main` package does the following : 

- first, define a bunch of tasks : 

```perl
task test => sub {...}
task test2 => sub {...}
...
task test6 => sub {...}
```
- Then, run each task and check that it has returned successfully (which constitutes the actual "test")

Just like in `t/needs.t`, some of those tasks invoke the `needs()` function, either with other tasks in the `main` package, or  tasks defined in other packages (`Rex::Module` or `Nested::Module`) which also reside in the same test script.

#### 1) Why do we have  the two other packages (`Rex::Module` and `Nested::Module`) ? 

Same reason as we have them in `t/needs.t`, i.e. we want to check that the `needs()` function works across package namespaces.

Reply:
#### 2) Why do we have an extra helper package (T), unlike in `t/needs.t` ?

Well, that's where the DRY refactoring happens... 

Notice how almost the same code is repeated again and again in the task subs in `t/needs.t`. 

Here, that boilerplate is replaced with calls to `track_taskrun` and `check_needed` which sensibly do the same, i.e. : 

- `track_taskrun`  :  When a task gets run, mark it as such by writing to a file with that name (and also save the its arguments there).

- `check_needed` : 
  * Make sure a "needed" task has already been run, i.e. : 
    * `die` unless a file with that name exists
    * `die` unless the arguments that were saved in that file correspond to the expected arguments
  * Also do some clean up, by deleting the said file. 
    Doing this here might not be considered very orthodox, but that's also what's happening in `t/needs.t`)

Notice how failure is signaled with `die`,  just like in `t/needs.t`. 

This also partly explains why we employ the private `_deep_check()`  routine from `Test::More::` here.

Just like you, my first reflex was employing `is_deeply`, but that doesn't work because "Test::More" chokes as it is then unable to keep track of the actual number of tests. 

The cleaner alternative would have been to rely on  `is_deeply` in the `main` test loop `for my $task ( $run_list->tasks )` in the test script's package `main`, but that won't work either, because the summary info (from `$task_list->get_summary`) does not contain task params/arguments... Well it probably should... :-)

Reply:
#### 3) Why do we have to do file IO in the first place ?

I was also suprised by this at first. I guess the author of  `t/needs.t` could have the best answer here. 

The explanation I came up with has to do with the task execution model in Rex (each task possibly having separate SSH connection, possibly running on a separate thread/process, ...).

A no-brainer method for eliminating the hairiness that would get involved is to go through file IO.

`t/needs.t` does this quite simply by writing/truncating a file that has the same name of the needed task when its being run, and then merely checking the existence of the file from the calling task.

Here, we also keep track of the params/arguments, so we use `Storable` to do that, which also eliviates the need for manually opening and reading/writing to files.

I chose `Storable` because of its simplicity and popularity. Besides, `Rex` already has a runtime dependency on `Storable`, so using it in a test did not entail adding a dependency.

Reply:
3) Why the unpacking logic ?

At one point you ask if we could get rid of the unpacking logic in helper routines, which was actually neatly placed on a single line (before perl tidy messed it up :-) :

```perl
    my %opts = ref $_[-1] eq 'HASH' ? %{; pop } : ();
```
Yes, we could; by directly assuming a HASH reference as the last argument, like such : 

```perl
    my %opts = %{; pop };
```
But, if you know of a way of preventing perl-tidy doing its thing, I think it would be preferable to keep the unpacking logic in there. 

It gives the caller more freedom. That way, if desired, the current tests in `t/needs.t` can easily be implemented with those two refactored routines (`track_taskrun` and `check_needed`). 

The above also gives away my thinking about you suggestion of merging these tests into `t/needs.t`. 

Imho, the easiest and most maintainable way of doing that would be retrofit `t/needs.t` tests into this -refactored- model.

Otherwise, if we try to it the non-DRY way `t/needs.t` does this (by copying and pasting the file IO code in all tasks), we could easily end up with a mess of several hundred lines (if not more) of hard-to-maintain code...

Reply:
> sure, please walk us through t/needs.t and I will do the same for this one :-)

Yeah, later I realized it is based on `t/needs.t`. My take is that `t/needs.t` is bad too, but that's part of the historical ballast from 10 years ago. That's a considerable amount of legacy to clean up, so that's why we use Test::Perl::Critic::Progressive to make sure we at least don't add more violations on top of the current pile :) These tests were unknowingly broken for PRs coming from forks, which was my mistake and I fixed that now (so it should properly complain about `t/issue/1508.t` after a rebase).

In other words, the policy is "we accept our legacy, but we don't want to add more ballast; therefore new code should be clean, or clean up old mess".

Reply:
Thanks for the detailed explanation!

---

>  1) Why do we have the two other packages (Rex::Module and Nested::Module) ?

Rex chops off the leading `Rex::` part for task names for brevity. We have these two modules defined in `t/needs.t` to make sure `needs` work with namespaces both with and without the `Rex::` prefix.

---

> 2) Why do we have an extra helper package (T), unlike in t/needs.t ?

> This also partly explains why we employ the private _deep_check() routine from Test::More:: here.
>
> Just like you, my first reflex was employing is_deeply, but that doesn't work because "Test::More" chokes as it is then unable to keep track of the actual number of tests.

That sounds like a good indicator to invent a cleaner approach than the 10-year-old legacy `t/needs.t` - or even better, fix `t/needs.t` first :)

I wonder if we could split `t/needs.t` up into real modules under e.g `t/lib` first as a pure refactor, then modify it to support running the same test tasks both with and without arguments :thinking: 

---

> 3) Why do we have to do file IO in the first place ?

Yup, I see we use files to keep track of some execution data, and I don't have a much better way for that currently. `Storable` also sounds good for dumping/restoring the extra data needed by the argument tests. It is also a Perl core module, so it wouldn't be a new dependency anyway :+1: 

---

> 3. Why the unpacking logic ?

> was actually neatly placed on a single line (before perl tidy messed it up :-)

That might also be an indicator of weirdness :) I'm happy to tune perltidy rules and reformat the codebase with that.

In this case though perlcritic complains about the null statement (and in the exploded formatting, about the missing final semicolon in the block). Perhaps it would be more readable as fully unpacking @_ first, and then do this transformation logic on the last element of the passed array?